# TVM Mortgage
<hr>
<p><i>A mortgage calculator that takes into account the time-value of money (TVM)</i></p>
<p>By Andrew Chap</p>
<hr>

## What do other mortgage calculators lack?
There are other good mortgage calculators out there, but they are limited to calculating monthly payments and the total nominal cost of a mortgage.  However, they don't differentiate a payment made now vs a payment made in the future.

## The time-value of money
<p>This calculator takes into account that money today is worth more than future money (assuming a positive interest rate).  To be specific, an amount of money today ${v_0}$ has a future value <span style = "color:#080#">$v_t$</span> at future time <span style = "color:#080#">$t$</span> (in years) by compounding at an (annual) interest rate <span style = "color:#080#">$i_\text{tvm}$</span> by the relation
</p>
<p style = "color:#080#">
    $$v_t = v_0(1+i_\text{tvm})^t.$$
</p>
<p>
    This is equivalent to saying that if you invest <span style = "color:#080#">$v_0$</span> in the market, which has an annual rate of return of <span style = "color:#080#">$i_\text{tvm}$</span>, after <span style = "color:#080#">$t$</span> years you can sell those investments for <span style = "color:#080#">$v_t$</span>.
</p>
<p>This also applies to payments made now vs. payments in the future.  If you are scheduled to make a payment of <span style = "color:#080#">$p_t$</span> in <span style = "color:#080#">$t$</span> years from now, the amount <span style = "color:#080#">$p_0$</span> you would need to invest today is
</p>
<p style = "color:#080#">
    $$p_0 = p_t(1+i_\text{tvm})^{-t}$$
</p>
<p>
which is simply a rearrangement of Eq. (1).  If you need to make a payment of <span style = "color:#080#">$p_t = \$100$</span> in <span style = "color:#080#">$t=5$</span> years, and the stock market has an <span style = "color:#080#">$i_\text{tvm}=6%$</span> annual return, an investment now of <span style = "color:#080#">$p_0 = \$75$</span> will suffice.  In other words, with an interest rate of <span style = "color:#080#">$i_\text{tvm}=6%$</span>, <span style = "color:#080#">\$100</span> in <span style = "color:#080#">5</span> years is worth <span style = "color:#080#">\$75</span> now.

## A demonstrative example
<p>"Paying points" is common option offered by mortgage lenders, in which you pay an upfront cost in order to lower your interest rate.  Say your mortgage lender offers you the following options on a
<span style="color:#080">
$l = \$100{,}000$</span>,
<span style="color:#080">
$t = 30\,\text{years}$
</span>
loan:</p>
<table>
    <tr align="center">
        <th>    &nbsp;                                </th>
        <th>    Upfront fee $$f$$                     </th>
        <th>    Mortgage interest rate $$i_\text{m}$$ </th>
        <th>    Monthly payment $$p$$                 </th>
    </tr>
    <tr>
        <td><i> Option 1                          </i></td>
        <td>    $\$0$                                 </td>
        <td>    $4.0\%$                               </td>
        <td>    $\$477.42$                            </td>
    </tr>
    <tr>
        <td><i> Option 2                          </i></td>
        <td>    $\$4{,}000$                           </td>
        <td>    $3.5\%$                               </td>
        <td>    $\$449.04$                            </td>
    </tr>
</table>
<p>where the monthly payment <span style="color:#080">$p$</span> is calculated from the standard payment formula:
</p>
<p style="color:#080">
$$p = l\frac{nc_\text{m}(1+c_\text{m})}{n(1 + c_\text{m})-1}$$
</p><p>
with 
</p><p style="color:#080">
$$
\begin{eqnarray}
n & \equiv & 30\,\text{years} \times 12\,\text{payments}/\text{year} = 360\,\text{payments} \\
c_\text{m} & \equiv & \frac{i_\text{m}}{12\,\text{payments}/\text{year}}
\end{eqnarray}
$$
</p><p>
and the total nominal cost is calculated by summing up all the payments plus any upfront costs:
</p><p style="color:#080">$$
\text{Total cost} = f + \sum_{n=0}^{360} p
$$</p>

### The nominal long-term cost of each option

<p>
With <i>Option 1</i> costing \$0 upfront with monthly payments of \$477.42 and <i>Option 2</i> costing \$4,000 upfront with monthly payments of \$449.04 (which would be referred to as paying for two "discount points" on your mortgage rate) it is easy to calculate the total nominal cost of each loan: 
</p>
<p style="color:#080"> $$
\begin{eqnarray}
\text{Option 1 total cost} & = & \$0     &+& \$477.42 \times 360 & = & \$171{,}870 \\
\text{Option 2 total cost} & = & \$4{,}000 &+& \$449.04 \times 360 & = & \$165{,}656
\end{eqnarray}
$$ </p>
Visually, we can plot the total amount you would pay for each loan as a function of time:

In [153]:
import plotly.offline as py; import plotly.graph_objs as go; import numpy as np
py.init_notebook_mode(connected=True)
r1 = 0.04/12.; r2 = 0.035/12.; time = np.linspace(0,30,361)
option1 = [100000*(r1*(1+r1)**360)/((1+r1)**360 - 1)*x for x in range(0,361)]
option2 = [5000 + 100000*(r2*(1+r2)**360)/((1+r2)**360 - 1)*x for x in range(0,361)]
data = [go.Scatter(x = time, y = option1, name = 'Option 1'),
        go.Scatter(x = time, y = option2, name = 'Option 2')]
layout = go.Layout(xaxis = {'title':'time (years)'}, yaxis = {'title':'Total spent ($)'})
fig = go.Figure(data=data,layout=layout); py.iplot(fig)

<p>
The intersection point happens at 11 years and 9 months.  I've had a lender tell me that this is how long it takes to recover the cost of paying for the discount points, and so if you own the house for longer than this length of time it is worth it to pay for the discount points if you can afford it.  If you keep the home for the entire length of the mortgage, you save more than \$6,000!  The flaw in this logic, of course, is that money today is worth more than the same quantity of money in the future, which we explore in the next section.
</p>

### Option 3: take option 1 but invest the \$4000 that you would have spent on Option 2
<p>
If you had the \$4,000 on hand to pay down the points, what if you chose option 1, invested that \$4,000, and each month you withdrew \$28.37 from your investment (the difference between the monthly payments of options 1 and 2) to go towards your monthly mortgage payment?  In that way you would still be paying \$4,000 upfront, and you would still need to come up with \$449.04/month for you mortgage payments, only instead of choosing option 2, you are choosing option 1 and paying your upfront cost into the stock market.
If the rate of return on your investment is 8% annually (which is equivalent to 0.49% compounded monthly), we can track your investment as follows.  In the first month you withdraw \$28.37, then your investment (called $I_\text{month 0})$ appreciates by 0.64%.
</p>
<p style="color:#080">
$$
I_\text{month 1} = \left(\$4{,}000 - \$28.37\right)\times(1 + 0.0064) = \$3997.18
$$
</p>
<p>
This iteration can be repeated through the 360<sup>th</sup> month:
</p>
<p style="color:#080">
$$ \begin{eqnarray}
I_\text{month 2} & = & \left(I_\text{month 1} - \$28.37\right)\times(1 + 0.0049) &=& \$3994.35 \\
I_\text{month 3} & = & \left(I_\text{month 2} - \$28.37\right)\times(1 + 0.0049) &=& \$3991.49 \\
 & ... & \\
I_\text{month 359} & = & \left(I_\text{month 358} - \$28.37\right)\times(1 + 0.0049) &=& \$60.25\\
I_\text{month 360} & = & \left(I_\text{month 359} - \$28.37\right)\times(1 + 0.0049) &=& \$32.08 
\end{eqnarray} $$
</p>
<p>
In this case, you've saved \$32.08 by choosing your "Option 3" version of Option 1, rather than Option 2.  Even better, if you sell your home before the end of the mortgage, you won't be losing out on any of the rate benefit you would have otherwise used the \$4,000 to pay for.
</p>
<p>
In other words, the time-value of money shifts the crossover point on how long you have to stay in the house to "make your money back" to justify paying for points, and in some cases (such as the one above) you never do.
</p>
<p>
In other words, if the market is performing at a rate of return of 8% (which is a reasonable long-term assumption) you are better off not paying for discount points <b> even if you think you will own the home for 30 years or more.</b>  Of course, you have to be able to tolerate the risk of investing that \$4,000, since the market fluctuates, and if you really did withdraw money from the market every month, the fees would eliminate any advantage.  Rather than being a real-world example, this is a demonstration of how the time value of money should influence mortgage decisions.
</p>

# Generalization
<p>
To visualize the advantage of option 1 over option 2 using the time-value of money, we can discount the future value of money by the interest rate.  To do this we define a future value discount <span style="color:#080">$\phi$</span> as a function of the month number <span style="color:#080">$n$</span>:
</p>
<p style="color:#080">
    $$ \phi(n) = (1 + c_\text{tvm})^{-n} $$
</p>
<p> where <span style="color:#080">$c_\text{tvm}$</span> is the monthly TVM (market) rate, calculated from the annual TVM rate <span style="color:#080">$i_\text{tvm}$</span> by:
<p style="color:#080">
    $$ c_\text{tvm} = (1 + i_\text{tvm})^\frac{1}{12} - 1 $$
</p><p>
which for <span style="color:#080">$i_\text{tvm} = 8\%$</span>, we get <span style="color:#080">$c_\text{tvm} = 0.64\%$</span>
</p><p>
We can then calculate the TVM-corrected total cost by multiplying each monthly payment by <span style="color:#080">$\phi$</span>:
<p style="color:#080">
    $$\text{Total cost} = f + p\sum_{n=0}^{360} (1 + c_\text{tvm})^{-n}$$
</p>
<p>
where <span style="color:#080">$f$</span> is the upfront cost of the mortgage and <span style="color:#080">$p$</span> is the monthly payment.  Using this equation, we can plot the TVM-corrected cost of each mortgage as follows:

In [294]:
import plotly.offline as py; import plotly.graph_objs as go; import numpy as np
py.init_notebook_mode(connected=True)
r1 = 0.04/12.; r2 = 0.035/12.; n=360; t=30; time = np.linspace(0,t,n+1)
i_tvm = 0.08
c_tvm = (1+i_tvm)**(1/12.) - 1
#c_tvm = i_tvm/12.
option1 = [0]*(n+1);
option2 = [0]*(n+1); option2[0] = 4000
for x in range(0,n):
    option1[x+1] = option1[x] + 100000*(r1*(1+r1)**n)/((1+r1)**n - 1)*(1+c_tvm)**(-x)
    option2[x+1] = option2[x] + 100000*(r2*(1+r2)**n)/((1+r2)**n - 1)*(1+c_tvm)**(-x)
data = [go.Scatter(x = time, y = option1, name = 'Option 1'),
        go.Scatter(x = time, y = option2, name = 'Option 2')]
layout = go.Layout(xaxis = {'title':'time (years)'}, yaxis = {'title':'Total spent ($)'})
fig = go.Figure(data=data,layout=layout); py.iplot(fig)
diff = (option2[-1] - option1[-1])
difftvm = diff*(1+c_tvm)**(n)
print(diff)
print(difftvm)
print("c_tvm = {}".format(c_tvm))

3.1878659551148303
32.078401314678615
c_tvm = 0.00643403011000343


In the case above, the difference in the total amount paid in option 1 and option 2 is only \$3.19, scaled by the TVM factor.  To find this difference in nominal dollars, the difference between the loan in TVM dollars is scaled by <span style="color:#080">$(1 + c_\text{tvm})^{360}$</span> to get \$32.08.  It is also interesting (and perhaps not too surprising) to note that even if you had the money on hand to buy a home in cash, as long as rate of return on the market (i.e. the TVM rate) is a bit higher than the mortgage interest rate, it is better to get a mortgage, and leaving otherwise available funds in the market.

In [291]:
#loan amount
l = 100000
#number of payments
n = 360

u1 = 0.
r1 = 0.04
u2 = 4000
r2 = 0.035
c1 = r1/12.
c2 = r2/12.

# principal+interest payment function
def pi(n,c,l):
    return l*(c*(1+c)**n)/((1+c)**n - 1.0)

p1 = pi(n,c1,l)
print("p1 = {}".format(p1))
p2 = pi(n,c2,l)
print("p2 = {}".format(p2))
diff = p1-p2
print("diff = {}".format(diff))

# invest upfront costs in the stockmarket, but withdrawl $diff each month
I = [None]*(n+1)
L = [None]*(n+1)
I[0] = u2
L[0] = l
i_tvm = 0.08
c_tvm = (1+i_tvm)**(1./12.) - 1.0
print("c_tvm = {}".format(c_tvm))

print("growth = {}".format(u2*(1+c_tvm)))
print("")

for i in range(1,n+1):
    I[i] = (I[i-1] - diff)*(1 + c_tvm)
    #I[i] = (I[i-1])*(1 + c_tvm) - diff
    L[i] = (L[i-1])*(1 + 0.04/12.) - p1
print(len(I))
print(I[-1])
print(I[-1]*(1+c_tvm)**(-n))

p1 = 477.4152954654538
p2 = 449.0446878088235
diff = 28.37060765663034
c_tvm = 0.00643403011000343
growth = 4025.7361204400136

361
32.07840131497622
3.1878659551444053


In [127]:
a = 0.04
areal = (1 + a/12.)**12. - 1
areal

0.040741542919790596

In [121]:
(2000 - 22.37)*(1.005)

1987.5181499999999

In [122]:
c

0.004867550565343048

In [284]:
I

[4000,
 3997.182975439481,
 3994.3478260581196,
 3991.494435240272,
 3988.622685619987,
 3985.7324590761764,
 3982.823636727758,
 3979.8960989287657,
 3976.9497252634264,
 3973.984394541209,
 3970.9999847918384,
 3967.99637326028,
 3964.973436401689,
 3961.931049876329,
 3958.8690885444585,
 3955.7874264611833,
 3952.6859368712753,
 3949.56449220396,
 3946.4229640676685,
 3943.261223244757,
 3940.07913968619,
 3936.876582506195,
 3933.653419976875,
 3930.409519522792,
 3927.1447477155134,
 3923.8589702681247,
 3920.5520520297046,
 3917.223856979767,
 3913.874248222666,
 3910.5030879819656,
 3907.1102375947708,
 3903.695557506026,
 3900.2589072627743,
 3896.80014550838,
 3893.3191299767145,
 3889.815717486305,
 3886.289763934444,
 3882.741124291264,
 3879.16965259377,
 3875.5752019398374,
 3871.9576244821687,
 3868.3167714222122,
 3864.652493004042,
 3860.9646385081974,
 3857.2530562454854,
 3853.517593550739,
 3849.7580967765402,
 3845.9744112868975,
 3842.166381450888,
 3838.333850636

In [226]:
L

[100000, 2.2992026060819626e-09]

In [68]:
r = 0.06
c = r/12.
print(c)
print((1+c)**12 - 1)

0.005
0.06167781186449828
